[gatsby-source-filesystem] add publicURL field

pieh opened the pull request:
This allow files to get copied to `static` directory and get their urls.

Closes #3538

KyleAMathews commented on packages/gatsby-source-filesystem/src/extend-file-node.js:
```diff
@@ -0,0 +1,36 @@
+      resolve: (file, fieldArgs, context) => {
+        const details = getNodeAndSavePathDependency(file.id, context.path)
+
+        const fileName = `${file.internal.contentDigest}${details.ext}`
```
Missing the `.` between the two strings?

KyleAMathews replied:
Also let's include the original file name in the new pathname like here: https://github.com/gatsbyjs/gatsby/blob/ce2f45cc27b8e0c065ca82ab82aca721c3720dd2/packages/gatsby-plugin-sharp/src/index.js#L265

pieh replied:
`details.ext` already contains `.`,  `details.extension` (or something like that) doesn't contain it.

Anyway this for this and rest of your definitly valid comments - this is basicly copied https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sharp/src/extend-node-type.js#L105 minus image width, height part, plus path prefix which is propably missing in code I linked, actual image processing methods (sizes, resolutions) do honor path prefix. 

I will fix those issues propably tomorrow.

KyleAMathews replied:
Oh haha, well please fix transformer-sharp while you're at it :-)

KyleAMathews commented on packages/gatsby-source-filesystem/src/extend-file-node.js:
```diff
@@ -0,0 +1,36 @@
+        if (!fs.existsSync(publicPath)) {
+          fs.copy(details.absolutePath, publicPath, err => {
+            if (err) {
+              console.error(`error copying file`, err)
```
Add the origin path and public path

KyleAMathews commented on packages/gatsby-source-filesystem/src/extend-file-node.js:
```diff
@@ -0,0 +1,36 @@
+        const publicPath = path.join(
+          process.cwd(),
+          `public`,
+          `static/${fileName}`
```
make `fileName` another argument instead of joining them here.

pieh commented:
Updated code. Also added gatsby-transformer-sharp changes here in seperate commit to not create more spam with seperate PR - if needed i can split it to keep it more contained.

KyleAMathews commented:
Thanks! Now I can finally stop asking people in discord to implement this to solve their problem 😅
